daemon/logger/loggerutils/logfile.go

```diff
@@ -224,6 +312,28 @@ func (w *LogFile) openRotatedFiles() (files []*os.File, err error) {
 			if !os.IsNotExist(err) {
 				return nil, err
 			}
+
+			tmpFile, err := decompressedJsonfile(fmt.Sprintf("%s.%d.gz", w.f.Name(), i-1), config.Since)
```

Comment:
This needs to be able to handle parallel readers. I don't think it will right now.
Maybe we can keep an index of decompressed files that are ref-counted and removed at 0 references?

Reply:
We may also want to optimize how we are opening files (and as such decompressing them) with regard to the number of lines requested with tail.

Reply:
>A good follow-up to this PR would be to to copy the file data in kernel space where available: https://github.com/moby/moby/blob/master/daemon/graphdriver/copy/copy.go

I think we could open the same file multiple times instead of copying the file (even hard links), only record the corresponding references-counted, looks like:
```go
    decompressedFileName := fmt.Sprintf("%s.%d.gz", w.f.Name(), i-1) + tmpLogfileSuffix
    var tmpFile *os.File
    // w.filesRefCounter  map[string]int
    if _, ok := w.filesRefCounter[decompressedFileName]; ok {
            w.filesRefCounter[decompressedFileName]++
            tmpFile, err = os.Open(decompressedFileName)
            if err != nil {
                    if !os.IsNotExist(err) {
                            return nil, err
                    }
                    continue
            }
    } else {
            tmpFile, err = decompressedJsonfile(fmt.Sprintf("%s.%d.gz", w.f.Name(), i-1), config.Since)
            if err != nil {
                    if !os.IsNotExist(err) {
                            return nil, err
                    }
                    continue
            }
            if tmpFile == nil {
                    // The log before `config.Since` does not need to read
                    break
            }
            w.filesRefCounter[tmpFile.Name()] = 1
    }

    files = append(files, tmpFile)
```
@cpuguy83 wdyt?

Reply:
More or less +1.
- The ref counter needs to be incremented immediately and not checked as this leads to racey behavior
- increment+open/create needs to be atomic
- decrement+remove should also be atomic
- Can probably keep a reference to the fd and use it multiple times instead of opening multiple times

Reply:
>Can probably keep a reference to the fd and use it multiple times instead of opening multiple times

I think it may be hard to use the same fd multiple times, because the offset of logfile may be changed when  logfile is read each time, this will cause confusion if a logfile is read by multiple readers at the same time.

Reply:
You're right we're using seek here. The implementation could be updated to use `ReaderAt` to allow this, but could be done separately.